source/Orders.cpp

```diff
@@ -0,0 +1,333 @@
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "Orders.h"
+
+#include "Ship.h"
+
+using namespace std;
+
+namespace {
+	using OrderSet = std::bitset<Orders::OrderType::TYPES_COUNT>;
+
+	// Bit mask to figure out which orders are canceled if their target
+	// ceases to be targetable or present.
+	const OrderSet REQUIRES_TARGET_SHIP(
+		(1 << Orders::OrderType::KEEP_STATION) +
+		(1 << Orders::OrderType::GATHER) +
+		(1 << Orders::OrderType::ATTACK) +
+		(1 << Orders::OrderType::FINISH_OFF));
+
+	const OrderSet REQUIRES_TARGET_ASTEROID(
+		(1 << Orders::OrderType::MINE));
+
+	// Bit masks that determine which orders may be given in conjunction with
+	// the key order. If an order isn't present in the mask of the key order
+	// then it will be set to false when the key order is given to a ship.
+	const map<Orders::OrderType, OrderSet> ORDER_MASKS = {
+		{Orders::HOLD_POSITION, OrderSet(0)},
+		{Orders::HOLD_ACTIVE, OrderSet(0)},
+		{Orders::MOVE_TO, OrderSet(0)},
+		{Orders::KEEP_STATION, OrderSet(0)},
+		{Orders::GATHER, OrderSet(0)},
+		{Orders::ATTACK, OrderSet(0)},
+		{Orders::FINISH_OFF, OrderSet(0)},
+		{Orders::MINE, OrderSet(0)},
+		{Orders::HARVEST, OrderSet(0)},
+	};
+}
+
+
+
+void Orders::SetHoldPosition()
+{
+	ApplyOrder(OrderType::HOLD_POSITION);
+}
+
+
+
+void Orders::SetHoldActive()
+{
+	ApplyOrder(OrderType::HOLD_ACTIVE);
+}
+
+
+
+void Orders::SetMoveTo()
+{
+	ApplyOrder(OrderType::MOVE_TO);
+}
+
+
+
+void Orders::SetKeepStation()
+{
+	ApplyOrder(OrderType::KEEP_STATION);
+}
+
+
+
+void Orders::SetGather()
+{
+	ApplyOrder(OrderType::GATHER);
+}
+
+
+
+void Orders::SetAttack()
+{
+	ApplyOrder(OrderType::ATTACK);
+}
+
+
+
+void Orders::SetFinishOff()
+{
+	ApplyOrder(OrderType::FINISH_OFF);
+}
+
+
+
+void Orders::SetMine()
+{
+	ApplyOrder(OrderType::MINE);
+}
+
+
+
+void Orders::SetHarvest()
+{
+	ApplyOrder(OrderType::HARVEST);
+}
+
+
+
+bool Orders::HasHoldPosition() const
+{
+	return activeOrders.test(OrderType::HOLD_POSITION);
+}
+
+
+
+bool Orders::HasHoldActive() const
+{
+	return activeOrders.test(OrderType::HOLD_ACTIVE);
+}
+
+
+
+bool Orders::HasMoveTo() const
+{
+	return activeOrders.test(OrderType::MOVE_TO);
+}
+
+
+
+bool Orders::HasKeepStation() const
+{
+	return activeOrders.test(OrderType::KEEP_STATION);
+}
+
+
+
+bool Orders::HasGather() const
+{
+	return activeOrders.test(OrderType::GATHER);
+}
+
+
+
+bool Orders::HasAttack() const
+{
+	return activeOrders.test(OrderType::ATTACK);
+}
+
+
+
+bool Orders::HasFinishOff() const
+{
+	return activeOrders.test(OrderType::FINISH_OFF);
+}
+
+
+
+bool Orders::HasMine() const
+{
+	return activeOrders.test(OrderType::MINE);
+}
+
+
+
+bool Orders::HasHarvest() const
+{
+	return activeOrders.test(OrderType::HARVEST);
+}
```

Comment:
It might be better to just have `Has(OrderType)` and `Set(OrderType)`.

Reply:
Perhaps, but I kinda wanted to avoid needing to have `Orders::OrderType::ORDER` all over the place in AI since that's a bit verbose.